File: rsocket-go-ffi/examples/simple-client.go
package main

import (
    "fmt"
    "log"
    "sync"
    "time"
    
    rsocket "../go"
)

func main() {
    fmt.Println("RSocket Go FFI Simple Client Example")
    fmt.Println("====================================")
    
    if err := rsocket.Init(); err != nil {
        log.Fatalf("Failed to initialize RSocket: %v", err)
    }
    
    fmt.Printf("RSocket version: %s\n", rsocket.GetVersion())
    fmt.Printf("Supported transports: %s\n", rsocket.GetSupportedTransports())
    
    client := rsocket.NewClient()
    if client == nil {
        log.Fatal("Failed to create client")
    }
    
    if err := client.ConnectTCP("127.0.0.1:7878"); err != nil {
        log.Fatalf("Failed to connect to server: %v", err)
    }
    
    fmt.Println("Connected to server")
    
    payload := rsocket.NewPayload("Hello from Go client!", "")
    if payload == nil {
        log.Fatal("Failed to create payload")
    }
    
    var wg sync.WaitGroup
    wg.Add(1)
    
    err := client.RequestResponse(payload, func(response *rsocket.Payload, err error) {
        defer wg.Done()
        
        if err != nil {
            fmt.Printf("Error: %v\n", err)
            return
        }
        
        if response != nil {
            fmt.Printf("Received response: %s\n", response.GetDataAsString())
        }
    })
    
    if err != nil {
        log.Fatalf("Failed to send request: %v", err)
    }
    
    wg.Wait()
    fmt.Println("Request completed")
}

File: rsocket-java-ffi/examples/SimpleClient.java
import com.rsocket.rust.*;
import java.util.concurrent.CompletableFuture;

public class SimpleClient {
    public static void main(String[] args) {
        System.out.println("RSocket Java FFI Simple Client Example");
        System.out.println("====================================");
        
        int result = RSocket.init();
        if (result != 0) {
            System.err.println("Failed to initialize RSocket");
            return;
        }
        
        System.out.println("RSocket version: " + RSocket.getVersion());
        
        try (RSocketClient client = new RSocketClient()) {
            client.connectTcp("127.0.0.1:7878");
            System.out.println("Connected to server");
            
            try (Payload payload = Payload.fromString("Hello from Java client!", "")) {
                CompletableFuture<Payload> future = client.requestResponse(payload);
                
                try (Payload response = future.get()) {
                    System.out.println("Received response: " + response.getDataAsString());
                }
            }
            
            System.out.println("Request completed");
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
